cpp/src/search/multisearch.cu

```diff
@@ -0,0 +1,604 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+#include "column_sequence_utils.hpp"
+#include "constexpr_equal_range.cuh"
+#include "index_range.hpp"
+
+#include <cudf.h>
+#include <utilities/cudf_utils.h>
+#include <utilities/bit_util.cuh>
+#include <utilities/error_utils.hpp>
+#include <utilities/column_utils.hpp>
+#include <utilities/integer_utils.hpp>
+#include <utilities/cuda_utils.hpp>
+#include <utilities/type_dispatcher.hpp>
+#include <search/fill.cuh>
+
+#include <rmm/thrust_rmm_allocator.h>
+
+#include <type_traits>
+
+#ifdef __NVCC__
+#define __fhd__ __forceinline__ __host__ __device__
+#define __fd__ __forceinline__ __device__
+#else
+#define __fhd__ inline
+#define __fd__ inline
+#endif
+
+template <typename I> __fhd__ constexpr I logical_xor (I x, I y) noexcept { return (    x and not y) or (not x and     y); }
+template <typename I> __fhd__ constexpr I logical_nxor(I x, I y) noexcept { return (    x and     y) or (not x and not y); }
+
+namespace warp {
+__fd__ auto ballot(int pred) {
+    enum : unsigned { full_warp_mask = 0xFFFFFFFFU };
+    return __ballot_sync(full_warp_mask, pred);
+}
+
+} // namespace warp
+
+namespace kernels {
+
+using index_range = cudf::util::index_range<gdf_size_type>;
+
+namespace thread {
+
+
+// TODO: Would be sped up/shortened by templating the flag.
+/**
+ * Find the subrange of NULL values within a contiguous subcolumn,
+ * provided that NULLs appear either all at the beginning or all
+ * at the end of the column.
+ *
+ * @param range We limit our search to the subcolumn this defines;
+ *     we cannot take iteraotrs since those are not defined for
+ *     bit-resolution data in a standard manner.
+ * @param validity_pseudocolumn The pseudocolumn of validity bits / null indicators;
+ *     we don't search all of it, only the designated @p range of indices
+ * @param nulls_appear_before_values If true, we may assume
+ *    @p validity_pseudocolumn begins with some NULLs (or none at all), and
+ *    at some point the NULLs end and only non-null values appear later.
+ *     When false, it's the same but for the _end_ of the column, i.e.
+ *     up to some point there are no nulls and then only nulls
+ *
+ * @note With a bit span class where the bit sequence does not
+ * necessarily begin at a bit container boundary, we could have
+ * just used @ref std::equal_range instead of this function.
+ *
+ * @return a subrange of @p range, whose values are all NULL (invalid),
+ * and with no NULLs outside it within @p range. It is guaranteed that
+ * if @p nulls_appear_before_values , the range will start at range.start_,
+ * and if not @p nulls_appear_before_values, the range will end at range.end_ -
+ * even for empty ranges.
+ */
+__device__ index_range
+nullness_subrange(
```

Comment:
I think this is overcomplicated. Could do it with less code and computation.

```
gdf_index_type *d_transition;
RMM_ALLOC(&d_transition, sizeof(gdf_index_type), stream);
CUDA_TRY(cudaMemsetAsync(transition, 0, stream));

thrust::for_each(rmm::exec_policy(stream)->on(stream),
                 thrust::make_counting_iterator(1), thrust::make_counting_iterator(n),
                 [validity, d_transition] __device__ (gdf_index_type i) { if (validity[i] != validity[i-1]) d_transition[i] = i; });

CUDA_TRY(cudaMemcpyAsync(h_transition, d_transition, sizeof(gdf_index_type), stream, cudaMemcpyDefault));
RMM_FREE(d_transition, stream);
```

Then just return the start and end of the range. You only need to call this once and then use this range in your binary searches. You don't need to compute the range of nulls in every thread's search.